src/configuration.rs

```diff
@@ -0,0 +1,128 @@
+//! Types for managing and extracting configuration.
+
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+
+use futures::future::FutureObj;
+
+use crate::{Extract, Request, Response, RouteMatch};
+
+trait ConfigurationItem: Any + Send + Sync {
+    fn clone_any(&self) -> Box<dyn ConfigurationItem>;
+    fn as_dyn_any(&self) -> &(dyn Any + Send + Sync);
+    fn as_dyn_any_mut(&mut self) -> &mut (dyn Any + Send + Sync);
+}
+
+impl<T> ConfigurationItem for T
+where
+    T: Any + Clone + Send + Sync,
+{
+    fn clone_any(&self) -> Box<dyn ConfigurationItem> {
+        Box::new(self.clone())
+    }
+
+    fn as_dyn_any(&self) -> &(dyn Any + Send + Sync) {
+        self
+    }
+
+    fn as_dyn_any_mut(&mut self) -> &mut (dyn Any + Send + Sync) {
+        self
+    }
+}
+
+impl Clone for Box<dyn ConfigurationItem> {
+    fn clone(&self) -> Box<dyn ConfigurationItem> {
+        (&**self).clone_any()
+    }
+}
+
+/// A cloneable typemap for saving per-endpoint configuration.
+///
+/// Configuration is mostly managed by `App` and `Router`, so this is normally not used directly.
+#[derive(Clone)]
```

Comment:
How should people debug this? If we're not adding the `Debug` trait, perhaps it would be useful to leave a note on how to inspect the values, or mention that it's currently not possible.

Reply:
Well, I think there are two options on implementing `Debug` for `Store`:
* Require `Debug` for all store items.
* Enable specialization, and specialize `Debug::fmt` for types that implements `Debug`.

First option works without any unstable features, but all items should implement `Debug`. Second one is unstable but potentially flexible as it doesn't require `Debug`. It seems like the former is better as deriving `Debug` is pretty trivial, but I'm not sure. How do you think, @yoshuawuyts?

Reply:
@tirr-c I think requiring `Debug` on all items is a reasonable constraint. In terms of semver it should probably be OK to loosen that constraint eventually (e.g. with specializations, maybe?), but we adding `Debug` as a constraint down the line would be a breaking change.

So I think it would be good to go with the first option, and we can always decide to loosen the constraint in the future.

Reply:
@yoshuawuyts I've added `Debug` impl for `Store`! It looks pretty good, check the `configuration` example.